src/client/Microsoft.Identity.Client/ManagedIdentity/ManagedIdentityKeyProviderFactory.cs

```diff
@@ -0,0 +1,42 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Threading;
+using Microsoft.Identity.Client.ManagedIdentity.Providers;
+using Microsoft.Identity.Client.PlatformsCommon.Shared; 
+
+namespace Microsoft.Identity.Client.ManagedIdentity
+{
+    internal static class ManagedIdentityKeyProviderFactory
+    {
+        private static IManagedIdentityKeyProvider s_provider;
+
+        internal static IManagedIdentityKeyProvider GetOrCreateProvider()
+        {
+            var p = Volatile.Read(ref s_provider);
+            if (p != null)
+                return p;
+
+            IManagedIdentityKeyProvider created = CreateProviderCore();
+            Interlocked.CompareExchange(ref s_provider, created, null);
+            return s_provider!;
```

Comment:
nit: we don't have nullable in msal, so `!` isn't needed?

Reply:
removed